internal/worker/worker.go

```diff
@@ -82,14 +82,28 @@ func (w *BaseWorker) Start(ctx context.Context) error {
 	dataChan := make(chan []byte)
 	readErrChan := make(chan error)
 
+	runCtx, runCancel := context.WithCancel(ctx)
+	defer runCancel()
+
 	go func() {
+		defer close(dataChan)
+		defer close(readErrChan)
+
 		for {
-			data, err := w.ioHandler.Read(ctx)
+			data, err := w.ioHandler.Read(runCtx)
```

Comment:
nit: can we move `w.ioHandler.Read` within the select statement and add it to the default case?
With this we may need a single select statement.

Reply:
This may not work since `Read` is blocking' and would cause the select case to stay blocked infinitely if the default case gets hit (and no new data comes by)